File: portal/backend/app/services/party_label.py
"""Format call parties as ``(Description) Extension`` for API responses."""

from __future__ import annotations


def _local_part(addr: str) -> str:
    return addr.split("@", 1)[0]


def party_extension(addr: str | None) -> str | None:
    if not addr:
        return None
    local = _local_part(addr).strip()
    if not local:
        return None
    digits = "".join(c for c in local if c.isdigit())
    if len(digits) >= 3:
        if len(digits) <= 7:
            return digits
        return local if local.startswith("+") else digits
    return local


def _strip_parens(value: str) -> str:
    trimmed = value.strip()
    if trimmed.startswith("(") and trimmed.endswith(")") and trimmed.count(")") == 1:
        return trimmed[1:-1].strip()
    return trimmed


def _already_formatted(name: str) -> bool:
    text = name.strip()
    if not (text.startswith("(") and ")" in text):
        return False
    rest = text.split(")", 1)[1].strip()
    return bool(rest)


def _comparable(value: str) -> str:
    digits = "".join(c for c in value if c.isdigit())
    return digits or value.strip().lower()


def format_party(name: str | None, addr: str | None) -> str:
    raw = (name or "").strip()
    if raw and _already_formatted(raw):
        return raw
    ext = party_extension(addr)
    label = _strip_parens(raw) if raw else ""
    if label and ext:
        if _comparable(label) == _comparable(ext):
            return ext
        return f"({label}) {ext}"
    if label:
        return label
    return ext or addr or "Unknown"

File: portal/frontend/src/utils/partyLabel.ts
/** Format a call party as `(Description) Extension` when both are known. */

function localPart(addr: string): string {
  return addr.includes('@') ? addr.split('@')[0] : addr;
}

/** Prefer a phone/extension id; fall back to SIP/email local part. */
export function partyExtension(addr?: string | null): string | null {
  if (!addr) return null;
  const local = localPart(addr).trim();
  if (!local) return null;
  const digits = local.replace(/\D/g, '');
  if (digits.length >= 3) {
    // Keep short extensions as-is; for longer DIDs prefer the raw local form
    // when it still looks numeric-ish (+E.164 etc.).
    if (digits.length <= 7) return digits;
    return local.startsWith('+') ? local : digits;
  }
  return local;
}

function stripWrappedParens(value: string): string {
  const trimmed = value.trim();
  if (trimmed.startsWith('(') && trimmed.endsWith(')') && trimmed.indexOf(')') === trimmed.length - 1) {
    return trimmed.slice(1, -1).trim();
  }
  return trimmed;
}

/** If UDS already stored `(Name) 1001`, reuse it as-is. */
function alreadyFormatted(name: string): boolean {
  return /^\(.+\)\s+\S+$/.test(name.trim());
}

function normalizeComparable(value: string): string {
  return value.replace(/\D/g, '').toLowerCase() || value.trim().toLowerCase();
}

/**
 * Display label for near/far parties.
 * Examples: `(Lamp, Michael) 4000`, `(Jeff Metcalf) jmetcalf`, `6026352608`
 */
export function formatParty(name?: string | null, addr?: string | null): string {
  const rawName = (name || '').trim();
  if (rawName && alreadyFormatted(rawName)) return rawName;

  const ext = partyExtension(addr);
  const label = rawName ? stripWrappedParens(rawName) : '';

  if (label && ext) {
    if (normalizeComparable(label) === normalizeComparable(ext)) return ext;
    return `(${label}) ${ext}`;
  }
  if (label) return label;
  return ext || addr || 'Unknown';
}
